drivers/tee/optee/smc_abi.c

```diff
@@ -907,6 +996,12 @@ static int optee_smc_do_call_with_arg(struct tee_context *ctx,
 	}
 	/* Initialize waiter */
 	optee_cq_wait_init(&optee->call_queue, &w);
+
+	/* May wait for a thread to become available in secure world */
+	if (!optee->thread.best_effort)
```

Comment:
This logic is too complicated, I'll try to propose something integrated with the `optee_cq_*()` functions.